test/interface/fast_chain.cpp

```diff
@@ -1,44 +1,1092 @@
+    database.update(block1, 1);
+    BOOST_REQUIRE_EQUAL(database.blocks().get(1, true).transaction_count(), 0);
+
+    // Setup ends.
+
+    const auto block = instance.get_candidate(1);
+    BOOST_REQUIRE(block == nullptr);
+}
+
+BOOST_AUTO_TEST_CASE(block_chain__get_candidate__present_with_transactions__success)
+{
+    START_BLOCKCHAIN(instance, false, true);
+    const auto bc_settings = bc::system::settings(config::settings::mainnet);
+    const chain::block& genesis = bc_settings.genesis_block;
+    auto& database = instance.database();
+    auto block1 = test::read_block(MAINNET_BLOCK1);
+    block1.set_transactions({ test::random_tx(0), test::random_tx(1) });
+    const auto incoming_headers = std::make_shared<const header_const_ptr_list>(header_const_ptr_list
+    {
+        std::make_shared<const message::header>(block1.header()),
+    });
+    const auto outgoing_headers = std::make_shared<header_const_ptr_list>();
+    BOOST_REQUIRE_EQUAL(database.reorganize({genesis.hash(), 0}, incoming_headers, outgoing_headers), error::success);
+    BOOST_REQUIRE(bc::database::is_candidate(database.blocks().get(1, true).state()));
+    database.update(block1, 1);
+    BOOST_REQUIRE_EQUAL(database.blocks().get(1, true).transaction_count(), 2);
+
+    // Setup ends.
+
+    const auto block = instance.get_candidate(1);
+    BOOST_REQUIRE(block->header().previous_block_hash() == genesis.hash());
+    BOOST_REQUIRE(block->header().merkle_root() != null_hash);
+    BOOST_REQUIRE(block->transactions().size() == 2);
+}
+
+BOOST_AUTO_TEST_CASE(block_chain__get_header__not_present__missing)
+{
+    START_BLOCKCHAIN(instance, false, true);
+
+    // Setup ends.
+
+    const auto block = instance.get_header(10, true);
+    BOOST_REQUIRE(block == nullptr);
+}
+
+// Writers.
+
+BOOST_AUTO_TEST_CASE(block_chain__store__no_state__failure)
+{
+    START_BLOCKCHAIN(instance, false, true);
+    auto& database = instance.database();
+    auto transaction = std::make_shared<const message::transaction>(test::random_tx(0));
+    BOOST_REQUIRE(!transaction->metadata.state);
+
+    // Setup ends.
+
+    BOOST_REQUIRE_EQUAL(instance.store(transaction), error::operation_failed);
+    BOOST_REQUIRE(!database.transactions().get(transaction->hash()));
+}
+
+BOOST_AUTO_TEST_CASE(block_chain__store__duplicate_transaction__failure)
+{
+    START_BLOCKCHAIN(instance, false, true);
+    const auto bc_settings = bc::system::settings(config::settings::mainnet);
+    const chain::block& genesis = bc_settings.genesis_block;
+    BOOST_REQUIRE_EQUAL(genesis.transactions().size(), 1);
+    auto transaction = std::make_shared<const message::transaction>(genesis.transactions()[0]);
+    transaction->metadata.state = instance.next_confirmed_state();
+
+    // Setup ends.
+
+#ifndef NDEBUG
+    BOOST_REQUIRE_EQUAL(instance.store(transaction), error::duplicate_transaction);
+#else
+    BOOST_REQUIRE(!instance.store(transaction));
+#endif
```

Comment:
Why do we end up with this `NDEBUG` condition??

Reply:
This is because the database can return a duplicate_transaction error without NDEBUG, and I think one of the travis builds runs test without NDEBUG. I had to do this in some test in database too.

https://github.com/libbitcoin/libbitcoin-database/blob/master/src/verify.cpp#L115